### PR TITLE
controllers: deploy csi correctly by adding labels

### PR DESCRIPTION
**Pull request by leelavg**
- fix the labels on deployment & daemonset resources
- add object meta to pod template spec for correct deployment
- create labels map if doesn't exist avoiding panic

there's a li'l pesky bug unrelated to current changeset as in the func createorupdate always updating the deployment & daemonset resources.

can be seen in the GH comments marked testing.

### CODE REVIEW

**Comment by leelavg** [comment]
Testing
- there is an unrelated bug you can see the log for createorupdate func is always showing operation as `updated`, might need to investigate if that needs to be fixed after we are done w/ all the refactoring and changing to config map as the owning resource.

```
===== CephFS deployment =====
# kc get deploy -lapp=csi-cephfsplugin-provisioner -ojsonpath={.items[*].spec.template.metadata}  | jq
{
  "creationTimestamp": null,
  "labels": {
    "app": "csi-cephfsplugin-provisioner"
  },
  "name": "csi-rbdplugin-provisioner",
  "namespace": "openshift-storage-client"
}

===== CephRBD deployment =====
# kc get deploy -lapp=csi-rbdplugin-provisioner -ojsonpath={.items[*].spec.template.metadata}  | jq
{
  "creationTimestamp": null,
  "labels": {
    "app": "csi-rbdplugin-provisioner"
  },
  "name": "csi-rbdplugin-provisioner",
  "namespace": "openshift-storage-client"
}

===== CephFS daemonset =====
# kc get ds -lapp=csi-cephfsplugin -ojsonpath={.items[*].spec.template.metadata}  | jq
{
  "creationTimestamp": null,
  "labels": {
    "app": "csi-cephfsplugin"
  },
  "name": "csi-cephfsplugin",
  "namespace": "openshift-storage-client"
}
 
===== CephRBD daemonset =====
# kc get ds -lapp=csi-rbdplugin -ojsonpath={.items[*].spec.template.metadata}  | jq
{
  "creationTimestamp": null,
  "labels": {
    "app": "csi-rbdplugin"
  },
  "name": "csi-rbdplugin",
  "namespace": "openshift-storage-client"
}

===== Logs from ocs-client-operator =====
[.....]

[Fresh deployment]

2024-03-15T05:45:20Z    INFO    selected sidecar images {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "2258bb73-d1cc-4e98-b5a7-cf8fdc2d8eaa", "ClusterVersion": {"name":"version"}, "version": "v4.16"}
2024-03-15T05:45:20Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "2258bb73-d1cc-4e98-b5a7-cf8fdc2d8eaa", "ClusterVersion": {"name":"version"}, "operation": "updated", "name": "ocs-csi-scc"}
2024-03-15T05:45:20Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "2258bb73-d1cc-4e98-b5a7-cf8fdc2d8eaa", "ClusterVersion": {"name":"version"}, "operation": "created", "name": "csi-cephfsplugin-provisioner"}
2024-03-15T05:45:20Z    INFO    Reconciling StorageClassClaim.  {"controller": "storageclassclaim", "controllerGroup": "ocs.openshift.io", "controllerKind": "StorageClassClaim", "StorageClassClaim": {"name":"ocs-storagecluster-cephfs"}, "namespace": "", "name": "ocs-storagecluster-cephfs", "reconcileID": "e85c2fc4-28ff-4c6c-b26e-6e81abf34f11", "StorageClassClaim": {"name":"ocs-storagecluster-cephfs"}}
2024-03-15T05:45:20Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "2258bb73-d1cc-4e98-b5a7-cf8fdc2d8eaa", "ClusterVersion": {"name":"version"}, "operation": "created", "name": "csi-cephfsplugin"}
2024-03-15T05:45:20Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "2258bb73-d1cc-4e98-b5a7-cf8fdc2d8eaa", "ClusterVersion": {"name":"version"}, "operation": "created", "name": "csi-rbdplugin-provisioner"}
2024-03-15T05:45:20Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "2258bb73-d1cc-4e98-b5a7-cf8fdc2d8eaa", "ClusterVersion": {"name":"version"}, "operation": "created", "name": "csi-rbdplugin"}

[Restart of ocs-client-operator]

2024-03-15T05:51:39Z    INFO    selected sidecar images {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "fc7829c5-048e-4619-b1a5-f778e899a274", "ClusterVersion": {"name":"version"}, "version": "v4.16"}
2024-03-15T05:51:40Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "fc7829c5-048e-4619-b1a5-f778e899a274", "ClusterVersion": {"name":"version"}, "operation": "updated", "name": "ocs-csi-scc"}
2024-03-15T05:51:40Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "fc7829c5-048e-4619-b1a5-f778e899a274", "ClusterVersion": {"name":"version"}, "operation": "updated", "name": "csi-cephfsplugin-provisioner"}
2024-03-15T05:51:40Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "fc7829c5-048e-4619-b1a5-f778e899a274", "ClusterVersion": {"name":"version"}, "operation": "updated", "name": "csi-cephfsplugin"}
2024-03-15T05:51:40Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "fc7829c5-048e-4619-b1a5-f778e899a274", "ClusterVersion": {"name":"version"}, "operation": "updated", "name": "csi-rbdplugin-provisioner"}
2024-03-15T05:51:40Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "fc7829c5-048e-4619-b1a5-f778e899a274", "ClusterVersion": {"name":"version"}, "operation": "updated", "name": "csi-rbdplugin"}
2024-03-15T05:51:40Z    INFO    successfully created or updated {"controller": "clusterversion", "controllerGroup": "config.openshift.io", "controllerKind": "ClusterVersion", "ClusterVersion": {"name":"version"}, "namespace": "", "name": "version", "reconcileID": "fc7829c5-048e-4619-b1a5-f778e899a274", "ClusterVersion": {"name":"version"}, "operation": "unchanged", "name": "prometheus-pvc-rules"}

[.....]
```

- w/o the fix, main is currently broken for deployments and it's better to have this at the earliest as to unblock others who rebased on main.

File: controllers/clusterversion_controller.go
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:218]: move this code to SetCephFSDeploymentDesiredState function. and also please take care of the same in other places also

File: pkg/csi/cephfsdaemonset.go
Comment (Madhu-1) [inline pkg/csi/cephfsdaemonset.go:51]: i haven't tried, all the labels present in the metadata need to be set in the selection, and here right? not just the defaults. Please check

File: pkg/csi/cephfsdeployment.go
Comment (Madhu-1) [inline pkg/csi/cephfsdeployment.go:193]: AFAIK This is not required and you are using `RBDDeploymentName` name here.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: controllers/clusterversion_controller.go
Comment (leelavg) [reply controllers/clusterversion_controller.go:218]: ya, initially I thought of doing it that way but wasn't sure a set function should create this or not.

File: pkg/csi/cephfsdaemonset.go
Comment (leelavg) [reply pkg/csi/cephfsdaemonset.go:51]: only one match is enough and selectors.matchLabels is immutable once set.

File: pkg/csi/cephfsdeployment.go
Comment (leelavg) [reply pkg/csi/cephfsdeployment.go:193]: > you are using RBDDeploymentName name here.
- good catch, thanks.

> This is not required
- we need to set pod name & ns isn't it?

Comment (Madhu-1) [reply pkg/csi/cephfsdeployment.go:193]: we need to set container names, pod names are auto generated from kubernetes based on the deployment name

Comment (leelavg) [reply pkg/csi/cephfsdeployment.go]: > pod names are auto generated
- I was just following previous behavior https://github.com/red-hat-storage/ocs-client-operator/blob/release-4.16/pkg/csi/cephfsdaemonset.go#L66-L68, do you want me to remove ns as well?

File: controllers/clusterversion_controller.go
Comment (leelavg) [reply controllers/clusterversion_controller.go]: done.

File: pkg/csi/cephfsdeployment.go
Comment (leelavg) [reply pkg/csi/cephfsdeployment.go]: done.

**Review by Madhu-1 — COMMENTED** [review]
LGTM

**Comment by Madhu-1** [comment]
/hold

for @nb-ohad review

**Review by Madhu-1 — APPROVED** [review]

File: pkg/csi/cephfsdaemonset.go
Comment (nb-ohad) [inline pkg/csi/cephfsdaemonset.go:262]: This pattern where the label map does not exist when trying to set it keeps appearing. 
Let's refactor into a util which returns guarantee labels on an object.
```
func AddLabels(obj metav1.Object, newLabels) {
	labels := obj.GetLabels()
	if labels == nil {
	         labels = map[string]string{}
                 obj.SetLabels(labels) 
	}
	maps.Copy(labels, newLabels)
}
```

Then we can simply do
```suggestion
	utils.AddLables(ds.Labels, cephfsDaemonsetLabels)
```

Comment (leelavg) [reply pkg/csi/cephfsdaemonset.go]: done.

File: pkg/utils/metadata.go
Comment (nb-ohad) [inline pkg/utils/metadata.go:1]: I see no reason for introducing a new file? Let's have this util in the k8sutils file

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: pkg/utils/metadata.go
Comment (leelavg) [reply pkg/utils/metadata.go:1]: I was told k8sutils are for those which takes kube client and so created new file. Nevertheless, fine w/ me.

Comment (leelavg) [reply pkg/utils/metadata.go:1]: done.

**Comment by leelavg** [comment]
> /hold
> 
> for @nb-ohad review

/unhold as Ohad gave the vote.

**Comment by leelavg** [comment]
> /hold
> 
> for @nb-ohad review

/unhold

as Ohad gave the vote.

**Pull request merged**
